diff -Naur evdi-1.9.1.old/module/evdi_drm_drv.c evdi-1.9.1/module/evdi_drm_drv.c
--- evdi-1.9.1.old/module/evdi_drm_drv.c	2021-04-13 18:01:36.248512202 +0200
+++ evdi-1.9.1/module/evdi_drm_drv.c	2021-04-13 18:01:50.239743359 +0200
@@ -79,7 +79,7 @@
 #endif
 
 static struct drm_driver driver = {
-#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE || defined(EL8)
 	.driver_features = DRIVER_MODESET | DRIVER_GEM | DRIVER_ATOMIC,
 #else
 	.driver_features = DRIVER_MODESET | DRIVER_GEM | DRIVER_PRIME
