diff -Naur evdi-1.13.0.old/module/evdi_drm_drv.h evdi-1.13.0/module/evdi_drm_drv.h
--- evdi-1.13.0.old/module/evdi_drm_drv.h	2023-03-06 13:05:18.000000000 +0100
+++ evdi-1.13.0/module/evdi_drm_drv.h	2023-03-17 07:24:22.478720989 +0100
@@ -28,7 +28,7 @@
 #endif
 #if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #include <drm/drm_framebuffer.h>
-#else
+#elif LINUX_VERSION_CODE <= KERNEL_VERSION (4, 18, 0-425) & defined(EL8)
 #include <drm/drm_irq.h>
 #endif
 #include <drm/drm_crtc.h>
diff -Naur evdi-1.13.0.old/module/evdi_gem.c evdi-1.13.0/module/evdi_gem.c
--- evdi-1.13.0.old/module/evdi_gem.c	2023-03-06 13:05:18.000000000 +0100
+++ evdi-1.13.0/module/evdi_gem.c	2023-03-17 07:24:22.478720989 +0100
@@ -10,10 +10,6 @@
 
 #include <linux/sched.h>
 #include <linux/version.h>
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
-#elif KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE
-#include <linux/dma-buf-map.h>
-#endif
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #include <drm/drm_prime.h>
 #include <drm/drm_file.h>
@@ -295,7 +291,7 @@
 	int ret;
 
 	if (evdi_drm_gem_object_use_import_attach(&obj->base)) {
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(4, 18, 0) <= LINUX_VERSION_CODE
 		struct iosys_map map = IOSYS_MAP_INIT_VADDR(NULL);
 #elif KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)
 		struct dma_buf_map map = DMA_BUF_MAP_INIT_VADDR(NULL);
@@ -328,7 +324,7 @@
 void evdi_gem_vunmap(struct evdi_gem_object *obj)
 {
 	if (evdi_drm_gem_object_use_import_attach(&obj->base)) {
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(4, 18, 0) <= LINUX_VERSION_CODE
 		struct iosys_map map = IOSYS_MAP_INIT_VADDR(NULL);
 
 		if (obj->vmap_is_iomem)
@@ -343,6 +339,7 @@
 
 		if (obj->vmap_is_iomem)
 			dma_buf_map_set_vaddr_iomem(&map, obj->vmapping);
+
 		else
 			dma_buf_map_set_vaddr(&map, obj->vmapping);
 
